### PR TITLE
Make streaming use the correct number of unique samples with SP/TP

### PR DESCRIPTION
**Pull request by snarayan21**
## Description of changes:
The TP/SP PR assumed that streaming would track/receive an increased number of samples when using SP/TP. For example, with `replication` = 2, Composer would previously track the # of elapsed samples without taking into account replication, which meant that even if 100 unique samples had been seen in training, since `replication` was 2, Composer would think that 200 samples had been seen, and pass that on to Streaming for resumption.

This issue was previously addressed by dividing the sample offset on resumption. With Composer now correctly calculating the number of unique samples elapsed during training, this fix is not needed on the Streaming side.

This change means we also need to track # unique samples using `StreamingDataLoader`.

Tested this with Shashank's help -- loss curves are overlapping on resumption.
Screenshot 2024-03-04 at 6 13 05 PM

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [ ] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [ ] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [ ] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [ ] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

File: streaming/base/dataloader.py [hunk 1]
```diff
@@ -73,6 +73,13 @@ def state_dict(self) -> Optional[Dict[str, Any]]:
         if isinstance(self.dataset, StreamingDataset):
             world = World.detect()
             num_samples = self.num_samples_yielded * world.num_ranks
+            if self.dataset.replication is not None:
+                # Check if we are using `replication`. If we are, then we need to adjust the
+                # `num_samples_yielded` to reflect the fact that sample ids are shared across
+                # `replication` consecutive devices. For example, if `replication` is 2, then the
+                # number of samples seen is half the number of samples yielded, since every pair
+                # of devices shares sample ids. So the index into the sample partition is halved.
+                num_samples = num_samples // self.dataset.replication
```
Comment (karan6181) [inline streaming/base/dataloader.py:82]: Why put this logic here and not in the dataset.py? Not everyone uses `StreamingDataloader`.

**Review by karan6181 — COMMENTED** [review]
Is there a way to valid this change using a unit test?

File: streaming/base/dataloader.py (hunk 1 above)
Comment (snarayan21) [reply streaming/base/dataloader.py:82]: So previously this logic was in `dataset.py`, but if someone is trying to get deterministic resumption, they should be using `StreamingDataLoader` anyways. The other path for people to get deterministic resumption is through Composer, which already takes care of this issue.

Comment (snarayan21) [reply streaming/base/dataloader.py:82]: `dataset.py` can't tell if the number of samples passed in is counting unique samples or not, so it's safer not to include it in the dataset.

**Comment by snarayan21** [comment]
@karan6181 it'll only come up during resumption, so Shashank and I tested it on a real run. Image is above

**Comment by karan6181** [comment]
> @karan6181 it'll only come up during resumption, so Shashank and I tested it on a real run. Image is above

Sync'd up offline. The plan is to add the test as part of integration testing.

**Review by karan6181 — APPROVED** [review]
